### PR TITLE
[SPLTRP-0862]: Extract reusable offline-first sync delegates to eliminate repository boilerplate

### LINKED ISSUE #862: Extract reusable offline-first sync delegates to eliminate repository boilerplate
## Context

The `:data` module has 5 entity repositories (`GroupRepositoryImpl`, `ExpenseRepositoryImpl`, `ContributionRepositoryImpl`, `CashWithdrawalRepositoryImpl`, `SubunitRepositoryImpl`) that share nearly identical offline-first coordination logic. This issue proposes extracting reusable functional delegates to eliminate the duplication while preserving the project's strict offline-first, real-time sync architecture.

### Inspiration

The idea of generic `networkBoundResource` (read delegate) and `offlineFirstMutation` (write delegate) inline functions was proposed as a starting point. However, after analyzing the actual codebase, the **proposed delegates do not directly fit** SplitTrip's architecture and need significant adaptation. See the gap analysis below.

---

### 1. Infrastructure Boilerplate (all 5 repos)

Every entity repository declares the same scaffolding:

```kotlin
private val syncScope = CoroutineScope(ioDispatcher)
private val cloudSubscriptionJobs = ConcurrentHashMap<String, Job>()  // or single Job? for Group
private val deletedPendingSyncIds: MutableSet<String> = ConcurrentHashMap.newKeySet()
```

### 2. Read Pattern: Flow + `onStart` Cloud Subscription (4 repos — Expense, Contribution, CashWithdrawal, Subunit)

All 4 keyed-by-group repositories use an identical structure:

```kotlin
override fun getGroupXxxFlow(groupId: String): Flow<List<Xxx>> =
    localDataSource.getXxxByGroupIdFlow(groupId)
        .onStart {
            cloudSubscriptionJobs[groupId]?.cancel()
            cloudSubscriptionJobs[groupId] = syncScope.launch {
                subscribeToCloudChanges(groupId)
            }
        }
```

`GroupRepositoryImpl` uses the same pattern but with a single `Job?` instead of `ConcurrentHashMap<String, Job>`.

### 3. `subscribeToCloudChanges()` — Reconciliation Logic (all 5 repos)

The reconciliation method is **near-identical** across all repositories (only the entity type and data source method names change):

```kotlin
private suspend fun subscribeToCloudChanges(groupId: String) {
    try {
        cloudDataSource.getXxxFlow(groupId)
            .collect { remoteItems ->
                try {
                    val filtered = if (deletedPendingSyncIds.isNotEmpty()) {
                        remoteItems.filter { it.id !in deletedPendingSyncIds }
                    } else {
                        remoteItems
                    }
                    localDataSource.replaceXxxForGroup(groupId, filtered)
                    if (deletedPendingSyncIds.isNotEmpty()) {
                        val remoteIds = remoteItems.map { it.id }.toSet()
                        deletedPendingSyncIds.removeAll(remoteIds)
                    }
                } catch (e: Exception) { Timber.w(e, "Error reconciling...") }
            }
    } catch (e: Exception) { Timber.w(e, "Error subscribing...") }
}
```

This is ~25 lines duplicated 5 times = **~125 lines** of near-copy-paste code.

### 4. Write Pattern: Create with Metadata + Background Sync (4 repos)

All 4 entity repositories follow the same create flow:

```kotlin
val id = entity.id.ifBlank { UUID.randomUUID().toString() }
val currentUserId = authenticationService.currentUserId() ?: ""
val currentTimestamp = LocalDateTime.now()
val entityWithMetadata = entity.copy(/* entity-specific fields */, syncStatus = SyncStatus.PENDING_SYNC)

localDataSource.saveXxx(entityWithMetadata)

syncScope.launch {
    try {
        cloudDataSource.addXxx(groupId, entityWithMetadata)
        localDataSource.updateSyncStatus(id, SyncStatus.SYNCED)
    } catch (e: Exception) {
        localDataSource.updateSyncStatus(id, SyncStatus.SYNC_FAILED)
    }
}
```

The structure is identical; only the entity-specific `.copy()` fields differ.

### 5. Delete Pattern with PENDING_SYNC Guard (4 repos)

All 4 entity repositories share the same delete guard:

```kotlin
val entity = localDataSource.getXxxById(id)
val wasPendingSync = entity?.syncStatus == SyncStatus.PENDING_SYNC
localDataSource.deleteXxx(id)
if (wasPendingSync) {
    deletedPendingSyncIds.add(id)
    return
}
syncScope.launch {
    try { cloudDataSource.deleteXxx(groupId, id) }
    catch (e: Exception) { Timber.w(e, "...") }
}
```

---

### `networkBoundResource` — Does NOT Fit

The proposed `networkBoundResource` follows the classic Android Architecture Components pattern (one-shot `fetchRemote` → `saveRemoteData` → observe local). **SplitTrip's reads are fundamentally different:**

| Aspect | Proposed `networkBoundResource` | SplitTrip's Actual Pattern |
|---|---|---|
| **Remote interaction** | One-shot `suspend` fetch | Continuous Firestore snapshot listener (`Flow`) |
| **Sync trigger** | `shouldFetch(localData)` check | Always subscribes; listener runs indefinitely |
| **Return type** | `Flow<Resource<T>>` (Loading/Success/Error) | `Flow<List<T>>` from Room directly |
| **UI contract** | UI handles `Resource` states | UI observes Room Flow; shimmer/empty states are separate |
| **Reconciliation** | Simple save | Merge strategy: upsert + selective delete + `deletedPendingSyncIds` filter |

Introducing `Resource<T>` would require changes to all repository interfaces, use cases, and ViewModels — a large surface-area change for questionable benefit given the real-time listener model.

### `offlineFirstMutation` — Partially Fits, Misses Key Patterns

The proposed write delegate captures the "save local → push remote" concept but:

1. **Missing `SyncStatus` transitions**: SplitTrip tracks `PENDING_SYNC → SYNCED / SYNC_FAILED` per entity. The proposed `markLocalSynced` callback is close but doesn't model the three-state transition or the current `updateSyncStatus(id, status)` pattern.
2. **Missing `deletedPendingSyncIds` guard for deletes**: The delete pattern has no equivalent in the proposed delegate.
3. **Entity-specific metadata enrichment** (UUID, timestamp, userId, entity-specific fields) varies per entity type and would still require per-repo lambdas.
4. **`GroupRepositoryImpl` has a two-phase verification** (`createGroup` → `verifyGroupOnServer`) and a `confirmPendingSyncGroups()` post-reconciliation step that don't fit the simple delegate.
5. **`enqueueForRetry`** is only used by `GroupRepositoryImpl` (via `GroupDeletionRetryScheduler`). Other repos currently mark as `SYNC_FAILED` without WorkManager retry.

---

## Proposed Solution: Adapted Delegates for SplitTrip

Instead of the generic proposed delegates, extract **SplitTrip-specific** functional utilities in a new `es.pedrazamiguez.splittrip.data.sync` package (under `:data`). These should be `inline` functions that compose with the existing data source interfaces.

### A. `RealtimeSyncCoordinator` — Read Delegate

Encapsulates the `localFlow.onStart { manageSubscription() }` + `subscribeToCloudChanges()` + `deletedPendingSyncIds` filtering pattern.

```kotlin
// Conceptual API — final design may differ
package es.pedrazamiguez.splittrip.data.sync

/**
 * Creates a Room-backed Flow that automatically manages a Firestore snapshot
 * subscription for real-time multi-device sync.
 *
 * Handles:
 * - Single-subscription enforcement (cancel before re-launch)
 * - deletedPendingSyncIds filtering (anti-resurrection)
 * - Merge reconciliation (upsert + selective delete)
 */
inline fun <T : Identifiable> realtimeSyncFlow(
    crossinline queryLocal: () -> Flow<List<T>>,
    crossinline observeRemote: () -> Flow<List<T>>,
    crossinline reconcileLocal: suspend (List<T>) -> Unit,
    deletedPendingSyncIds: MutableSet<String>,
    subscriptionTracker: SubscriptionTracker,
    syncScope: CoroutineScope,
    entityName: String  // for Timber logging
): Flow<List<T>>
```

**Two `SubscriptionTracker` variants:**
- `SingleSubscriptionTracker` — for `GroupRepositoryImpl` (single `Job?`)
- `KeyedSubscriptionTracker` — for per-group repos (`ConcurrentHashMap<String, Job>`)

### B. `offlineFirstCreate` — Write Delegate (Create)

Encapsulates the "enrich metadata → save local (PENDING_SYNC) → background sync → update status" pattern.

```kotlin
suspend inline fun <T> offlineFirstCreate(
    crossinline saveLocal: suspend () -> T,
    crossinline pushRemote: suspend (T) -> Unit,
    crossinline onSynced: suspend (T) -> Unit,       // updateSyncStatus(id, SYNCED)
    crossinline onSyncFailed: suspend (T) -> Unit,    // updateSyncStatus(id, SYNC_FAILED)
    syncScope: CoroutineScope,
    entityName: String
): T
```

> **Note:** Entity-specific metadata enrichment (`.copy(id = uuid, createdBy = ..., ...)`) remains in the repository as it's inherently entity-specific. The delegate only encapsulates the local-first + background-sync orchestration.

### C. `offlineFirstDelete` — Write Delegate (Delete)

Encapsulates the "check PENDING_SYNC → delete local → guard or sync" pattern.

```kotlin
suspend inline fun offlineFirstDelete(
    crossinline getSyncStatus: suspend () -> SyncStatus?,
    crossinline deleteLocal: suspend () -> Unit,
    crossinline pushRemoteDeletion: suspend () -> Unit,
    crossinline onRemoteFailed: suspend (Exception) -> Unit,
    deletedPendingSyncIds: MutableSet<String>,
    entityId: String,
    syncScope: CoroutineScope,
    entityName: String
)
```

### D. `offlineFirstUpdate` — Write Delegate (Update)

Simpler variant of `offlineFirstCreate` for updates (no UUID generation, no PENDING_SYNC guard on failure — just overwrite locally and sync).

---

## What Remains Entity-Specific (NOT Extractable)

These concerns are intentionally **not** part of the delegates:

1. **Metadata enrichment** in `.copy()` calls (entity-specific fields like `members`, `withdrawnBy`, `remainingAmount`)
2. **`GroupRepositoryImpl`'s two-phase verification** (`createGroup` → `verifyGroupOnServer`)
3. **`GroupRepositoryImpl`'s `confirmPendingSyncGroups()`** post-reconciliation hook (see #861)
4. **`GroupRepositoryImpl`'s `GroupDeletionRetryScheduler`** integration
5. **`CashWithdrawalRepositoryImpl`'s `updateRemainingAmounts()`** batch pattern
6. **`ContributionRepositoryImpl`'s `deleteByLinkedExpenseId()`** linked deletion

The delegates reduce the **common scaffolding**; the entity-specific business logic stays in each repository.

---

## Estimated Impact

| Metric | Before | After (estimated) |
|---|---|---|
| `subscribeToCloudChanges()` copies | 5 × ~25 lines | 1 reusable function |
| `getGroupXxxFlow()` + `onStart` copies | 5 × ~8 lines | 1 reusable function |
| Create boilerplate per repo | ~20 lines | ~5 lines (lambda config only) |
| Delete boilerplate per repo | ~15 lines | ~3 lines (lambda config only) |
| Infrastructure declarations per repo | ~5 lines | Encapsulated in coordinator |
| **Total boilerplate reduced** | **~300+ lines across 5 repos** | **Centralized in ~100-line utility module** |

---

### Phase 1: Core Delegates
1. Create `data/src/main/kotlin/.../data/sync/` package
2. Implement `Identifiable` interface (or use a lambda `(T) -> String` for ID extraction)
3. Implement `SubscriptionTracker` (single + keyed variants)
4. Implement `realtimeSyncFlow()` inline function
5. Implement `offlineFirstCreate()`, `offlineFirstDelete()`, `offlineFirstUpdate()`
6. Unit tests for each delegate using `StandardTestDispatcher` injection

### Phase 2: Migrate Repositories (one at a time)
1. `SubunitRepositoryImpl` — simplest, good first candidate
2. `ContributionRepositoryImpl` — similar complexity, validates pattern
3. `ExpenseRepositoryImpl` — slightly more complex
4. `CashWithdrawalRepositoryImpl` — has `updateRemainingAmounts` batch edge case
5. `GroupRepositoryImpl` — most complex, has two-phase verification and WorkManager retry

### Phase 3: Validation
1. Ensure all existing repository unit tests pass without modification (behavior must be identical)
2. Run CPD to confirm duplication is eliminated
3. Run Konsist architecture tests to ensure no new violations

---

## Acceptance Criteria

- [ ] New `data/sync/` package with reusable inline utility functions
- [ ] `realtimeSyncFlow()` encapsulates: local Flow observation, single-subscription management, `deletedPendingSyncIds` filtering, merge reconciliation, and Timber logging
- [ ] `offlineFirstCreate()` encapsulates: local save (PENDING_SYNC) → background cloud push → `SyncStatus` transition (SYNCED / SYNC_FAILED)
- [ ] `offlineFirstDelete()` encapsulates: PENDING_SYNC guard → local delete → background cloud delete → `deletedPendingSyncIds` tracking
- [ ] `offlineFirstUpdate()` encapsulates: local save → background cloud push → `SyncStatus` transition
- [ ] All 5 entity repositories are migrated to use the delegates
- [ ] `GroupRepositoryImpl`'s unique patterns (two-phase verification, `confirmPendingSyncGroups`, `GroupDeletionRetryScheduler`) are preserved via hook lambdas or remain inline
- [ ] All existing repository unit tests pass without changes (behavior-preserving refactor)
- [ ] No `Resource<T>` sealed class introduced — the UI contract (`Flow<List<T>>` from Room) remains unchanged
- [ ] CPD report shows reduced duplication in `:data` module
- [ ] No new Detekt findings introduced
- [ ] `CoroutineDispatcher` injection is preserved for testability

## Related Issues

- #861 — Add `confirmPendingSync` mechanism to non-group repositories (the post-reconciliation hook could be a parameter of `realtimeSyncFlow` once this is implemented)

### PR DESCRIPTION
**Pull request by pedrazamiguez**
## 🔗 Related Issues

Closes #862 

## 🧪 Testing Instructions

- [ ] I have tested these changes locally
- [ ] I have added/updated tests as needed
- [ ] All existing tests pass

## ✅ Quality Checklist

- [ ] My code follows the project's coding standards
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have updated documentation as needed

### CODE REVIEW

File: data/src/test/kotlin/es/pedrazamiguez/splittrip/data/sync/CloudSyncDelegatesTest.kt [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+        fun `continues collecting after reconciliation error`() = runTest(testDispatcher) {
+            // The delegate catches inner exceptions — the outer flow keeps collecting.
+            // With flowOf, there's only one emission, so we just verify it doesn't throw.
+            var reconcileCallCount = 0
+
+            subscribeAndReconcile(
+                cloudFlow = flowOf(listOf("item-1")),
+                reconcileLocal = {
+                    reconcileCallCount++
+                    throw RuntimeException("Reconciliation error")
```
[the remark this answers is not in the document]
Comment (pedrazamiguez) [reply data/src/test/kotlin/es/pedrazamiguez/splittrip/data/sync/CloudSyncDelegatesTest.kt]: ✅ Valid. The project convention in test files is to use `IOException` for simulating network/error failures (see `GroupRepositoryImplTest`, `ExpenseRepositoryImplTest`, etc.). Replacing all 5 occurrences with `IOException`.

File: data/src/test/kotlin/es/pedrazamiguez/splittrip/data/sync/CloudSyncDelegatesTest.kt [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+                assertEquals(emptyList<String>(), syncedIds)
+            }
+
+        @Test
+        fun `does not mark items when server verification throws`() = runTest(testDispatcher) {
+            val syncedIds = mutableListOf<String>()
+
+            confirmPendingSync(
+                getPendingIds = { listOf("id-1") },
+                verifyOnServer = { throw RuntimeException("Server unreachable") },
```
[the remark this answers is not in the document]
Comment (pedrazamiguez) [reply data/src/test/kotlin/es/pedrazamiguez/splittrip/data/sync/CloudSyncDelegatesTest.kt]: Same fix as above — replacing with `IOException`.

File: data/src/test/kotlin/es/pedrazamiguez/splittrip/data/sync/CloudSyncDelegatesTest.kt [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+        @Test
+        fun `continues to next item when one verification fails`() = runTest(testDispatcher) {
+            val syncedIds = mutableListOf<String>()
+            var callCount = 0
+
+            confirmPendingSync(
+                getPendingIds = { listOf("id-1", "id-2", "id-3") },
+                verifyOnServer = { id ->
+                    callCount++
+                    if (id == "id-2") throw RuntimeException("Error")
```
[the remark this answers is not in the document]
Comment (pedrazamiguez) [reply data/src/test/kotlin/es/pedrazamiguez/splittrip/data/sync/CloudSyncDelegatesTest.kt]: Same fix as above — replacing with `IOException`.

File: data/src/test/kotlin/es/pedrazamiguez/splittrip/data/sync/CloudSyncDelegatesTest.kt [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+        }
+
+        @Test
+        fun `marks SYNC_FAILED when cloud write throws`() = runTest(testDispatcher) {
+            val updateSyncStatus: suspend (String, SyncStatus) -> Unit = mockk(relaxed = true)
+
+            syncCreateToCloud(
+                scope = this,
+                entityId = "entity-1",
+                cloudWrite = { throw RuntimeException("Network error") },
```
[the remark this answers is not in the document]
Comment (pedrazamiguez) [reply data/src/test/kotlin/es/pedrazamiguez/splittrip/data/sync/CloudSyncDelegatesTest.kt]: Same fix as above — replacing with `IOException`.

File: data/src/test/kotlin/es/pedrazamiguez/splittrip/data/sync/CloudSyncDelegatesTest.kt [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+            coVerify(exactly = 1) { cloudDelete() }
+        }
+
+        @Test
+        fun `does not throw when cloud delete fails`() = runTest(testDispatcher) {
+            // Should not propagate the exception
+            syncDeletionToCloud(
+                scope = this,
+                entityId = "entity-1",
+                cloudDelete = { throw RuntimeException("Network error") },
```
[the remark this answers is not in the document]
Comment (pedrazamiguez) [reply data/src/test/kotlin/es/pedrazamiguez/splittrip/data/sync/CloudSyncDelegatesTest.kt]: Same fix as above — replacing with `IOException`.

File: data/src/main/kotlin/es/pedrazamiguez/splittrip/data/sync/KeyedSubscriptionTracker.kt [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+     * @param scope CoroutineScope to launch the subscription in.
+     * @param block Suspend function to run as the subscription body.
+     */
+    fun cancelAndRelaunch(
+        key: String,
+        scope: CoroutineScope,
+        block: suspend () -> Unit
+    ) {
+        jobs[key]?.cancel()
+        jobs[key] = scope.launch { block() }
```
[the remark this answers is not in the document]
Comment (pedrazamiguez) [reply data/src/main/kotlin/es/pedrazamiguez/splittrip/data/sync/KeyedSubscriptionTracker.kt]: ✅ Good catch. Adopting `compute()` for atomic cancel + replace. This eliminates the narrow race window between the read and write.

File: data/src/main/kotlin/es/pedrazamiguez/splittrip/data/sync/CloudSyncDelegates.kt [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+                Timber.d("Real-time sync: %d %ss %s", remoteItems.size, entityLabel, logContext)
+                reconcileLocal(remoteItems)
+                confirmPendingSync(getPendingIds, verifyOnServer, markSynced, entityLabel)
+            } catch (e: Exception) {
+                Timber.w(e, "Error reconciling %ss from cloud snapshot", entityLabel)
+            }
+        }
+    } catch (e: Exception) {
+        Timber.w(e, "Error subscribing to cloud %s changes, using local cache", entityLabel)
+    }
```
[the remark this answers is not in the document]
Comment (pedrazamiguez) [reply data/src/main/kotlin/es/pedrazamiguez/splittrip/data/sync/CloudSyncDelegates.kt:58]: ✅ Valid. Adding `CancellationException` rethrow in both the inner and outer `catch` blocks. This ensures coroutine cancellation (e.g., subscription relaunch via `KeyedSubscriptionTracker`) propagates correctly instead of being logged as an error.

File: data/src/main/kotlin/es/pedrazamiguez/splittrip/data/sync/CloudSyncDelegates.kt [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+) {
+    scope.launch {
+        try {
+            cloudWrite()
+            updateSyncStatus(entityId, SyncStatus.SYNCED)
+            Timber.d("%s synced to cloud: %s", entityLabel.replaceFirstChar { it.uppercase() }, entityId)
+        } catch (e: Exception) {
+            updateSyncStatus(entityId, SyncStatus.SYNC_FAILED)
+            Timber.w(e, "Failed to sync %s to cloud", entityLabel)
+        }
```
[the remark this answers is not in the document]
Comment (pedrazamiguez) [reply data/src/main/kotlin/es/pedrazamiguez/splittrip/data/sync/CloudSyncDelegates.kt:132]: ✅ Valid. Adding `CancellationException` rethrow so cancellation doesn't incorrectly mark entities as `SYNC_FAILED`.

File: data/src/main/kotlin/es/pedrazamiguez/splittrip/data/sync/CloudSyncDelegates.kt [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+    for (id in pendingIds) {
+        try {
+            if (verifyOnServer(id)) {
+                markSynced(id)
+                Timber.d("Confirmed %s sync: %s", entityLabel, id)
+            }
+        } catch (e: Exception) {
+            Timber.d(e, "Cannot confirm %s %s — server unreachable", entityLabel, id)
+        }
+    }
```
[the remark this answers is not in the document]
Comment (pedrazamiguez) [reply data/src/main/kotlin/es/pedrazamiguez/splittrip/data/sync/CloudSyncDelegates.kt:97]: ✅ Valid. Adding `CancellationException` rethrow so the for-loop stops promptly when the parent job is cancelled instead of continuing to the next item.
